packages/controller-utils/src/siwe.ts

```diff
@@ -34,10 +34,141 @@ function msgHexToText(hex: string): string {
   }
 }
 
+/**
+ * @type WrappedSIWERequest
+ *
+ * Sign-In With Ethereum (SIWE)(EIP-4361) message with request metadata
+ * @property {string} from - Subject account address
+ * @property {string} origin - The RFC 3986 originating authority of the signing request, including scheme
+ * @property {ParsedMessage} siwe - The data parsed from the message
+ */
+export interface WrappedSIWERequest {
+  from: string;
+  origin: string;
+  siwe: SIWEMessage;
+}
+
+interface HostParts extends DomainParts {
+  scheme: string;
+}
+
+interface DomainParts {
+  userinfo: string | null;
+  host: string;
+  port: number | null;
+}
+
+export const ORIGIN_REGEX =
+  /^(?<scheme>[^:/]+):\/\/((?<userinfo>[^@]*)@)?(?<host>[^:@/?]+)(:(?<port>[0-9]+))?/u;
+export const DOMAIN_REGEX =
+  /^((?<userinfo>[^@]*)@)?(?<host>[^:@/?]+)(:(?<port>[0-9]+))?$/u;
+const DEFAULT_PORTS_BY_SCHEME = {
+  http: 80,
+  https: 443,
+} as { [key: string]: number | undefined };
+
+/**
+ * Parses parts from RFC 3986 authority from environment relevant for EIP-4361 origin validation.
+ *
+ * @param origin - input string
+ * @returns parsed parts
+ */
+export const parseOriginParts = (origin: string): HostParts => {
+  const match = origin.match(ORIGIN_REGEX);
+  if (!match) {
+    throw new Error(`Unrecognized origin format "${origin}".`);
+  }
+  const { scheme, userinfo, host, port } = match.groups as {
+    scheme: string;
+    userinfo: string | undefined;
+    host: string;
+    port: string | undefined;
+  };
+  return {
+    scheme,
+    // strip password from userinfo
+    userinfo: userinfo ? userinfo.replace(/:.*$/u, '') : null,
+    host,
+    port: port
+      ? parseInt(port)
+      : DEFAULT_PORTS_BY_SCHEME[scheme.toLocaleLowerCase()] || null,
+  };
+};
+
+/**
+ * Parses parts from RFC 3986 authority from EIP-4361 `domain` field.
+ *
+ * @param domain - input string
+ * @returns parsed parts
+ */
+export const parseDomainParts = (domain: string): DomainParts => {
```

Comment:
Custom parsing for domain makes more sense, so that we know if port was there or default was assumed. I'd still add a scheme to the string and pass it through URL and see if I'm getting matching result